### ISSUE TITLE
Does randomized TruncatedSVD really work on big, non-skinny datasets?

### ISSUE THREAD

**Issue by georgeblck** [dask/dask-ml#604]
I am trying to calculate a TruncatedSVD on data that is too big to fit in memory. The original `sklearn.decomposition` implementation does not work in any case, so I have been trying to use the dask-ml version (that is based on the `dask.array.linalg.svd_compressed` algorithm).

Here is a minimal code that uses data similar to mine and gets killed because of too much RAM; depending on your locally available RAM the dimensions of the dataset might have to be adapted.

```
import dask_ml.decomposition ## version 1.2.0
import dask.array as da  ## version 2.9.2
import numpy as np

# Create Random data
bigData = da.random.random((40000, 500000)).astype(np.float32).rechunk('auto')
# Fit the SVD
svd = sklearn.decomposition.TruncatedSVD(4000, "randomized")
svd.fit(test)
smallDat = svd.transform(test)
```

Maybe this is connected to  #401 or I just don't understand the use cases of dask? Or maybe I am chunking my data wrong?

Thanks in advance for any help!

**Comment by TomAugspurger** [comment]
It's intended for tall and skinny datasets:
https://docs.dask.org/en/latest/array-api.html#dask.array.linalg.tsqr

**Comment by georgeblck** [comment]
Yes, the default algorithm is "tsqr" which is only supported for tall and skinny datasets. However, the "randomized" version is based on the approximate algorithm based on `dask.array.linalg.svd_compressed` also referenced in the [dask examples](https://examples.dask.org/machine-learning/svd.html).

So either the `svd_compressed` does not work with that big data (I will test it), or there is a problem in how TruncatedSVD handles the data?

**Comment by TomAugspurger** [comment]
I’m not sure.

**Comment by mrocklin** [comment]
The compressed solution can work on non-skinny datasets, yes.  If your data
does not fit in memory I recommend using the compute=True option.  You will
also probably have to be mindful of the chunking of the data in order for
the matrix multiplies to fit nicely in small memory (this is hard).  I
started writing a blogpost on this topic here but never finished:
https://github.com/dask/dask-blog/pull/38/files

**Comment by georgeblck** [comment]
Thank you @mrocklin , that was exactly the information I needed. With the `compute=True` option, the `dask.array.linalg.svd_compressed` works on this huge array. Although it does take a while, but that is not the problem.
Regarding the chunking, I could still use the `rechunk('auto')` option with `dask.config.set({'array.chunk-size': '1GiB'})`. So no manual chunk-size specification needed.

The only issue now is that the `TruncatedSVD`-wrapper does not work, because the compute-option cannot be passed. But for me this is not so important.

Thanks again for the quick help.

**Comment by mrocklin** [comment]
FYI Auto rechunking is unlikely to give optimal performance here.  I
encourage you to play around a bit.

**Comment by mrocklin** [comment]
(I'm glad that things are working for you)
